stl/inc/numbers

```diff
@@ -0,0 +1,188 @@
+    template <>
+    inline constexpr double phi_v<double> = static_cast<double>(1.618033988749895);
+
+    template <>
+    inline constexpr float e_v<float> = static_cast<float>(e_v<double>);
+    template <>
+    inline constexpr float log2e_v<float> = static_cast<float>(log2e_v<double>);
+    template <>
+    inline constexpr float log10e_v<float> = static_cast<float>(log10e_v<double>);
+    template <>
+    inline constexpr float pi_v<float> = static_cast<float>(pi_v<double>);
+    template <>
+    inline constexpr float inv_pi_v<float> = static_cast<float>(inv_pi_v<double>);
+    template <>
+    inline constexpr float inv_sqrtpi_v<float> = static_cast<float>(inv_sqrtpi_v<double>);
+    template <>
+    inline constexpr float ln2_v<float> = static_cast<float>(ln2_v<double>);
+    template <>
+    inline constexpr float ln10_v<float> = static_cast<float>(ln10_v<double>);
+    template <>
+    inline constexpr float sqrt2_v<float> = static_cast<float>(sqrt2_v<double>);
+    template <>
+    inline constexpr float sqrt3_v<float> = static_cast<float>(sqrt3_v<double>);
+    template <>
+    inline constexpr float inv_sqrt3_v<float> = static_cast<float>(inv_sqrt3_v<double>);
+    template <>
+    inline constexpr float egamma_v<float> = static_cast<float>(egamma_v<double>);
+    template <>
+    inline constexpr float phi_v<float> = static_cast<float>(phi_v<double>);
+
+    template <>
+    inline constexpr long double e_v<long double> = static_cast<long double>(e_v<double>);
+    template <>
+    inline constexpr long double log2e_v<long double> = static_cast<long double>(log2e_v<double>);
+    template <>
+    inline constexpr long double log10e_v<long double> = static_cast<long double>(log10e_v<double>);
+    template <>
+    inline constexpr long double pi_v<long double> = static_cast<long double>(pi_v<double>);
+    template <>
+    inline constexpr long double inv_pi_v<long double> = static_cast<long double>(inv_pi_v<double>);
+    template <>
+    inline constexpr long double inv_sqrtpi_v<long double> = static_cast<long double>(inv_sqrtpi_v<double>);
+    template <>
+    inline constexpr long double ln2_v<long double> = static_cast<long double>(ln2_v<double>);
+    template <>
+    inline constexpr long double ln10_v<long double> = static_cast<long double>(ln10_v<double>);
+    template <>
+    inline constexpr long double sqrt2_v<long double> = static_cast<long double>(sqrt2_v<double>);
+    template <>
+    inline constexpr long double sqrt3_v<long double> = static_cast<long double>(sqrt3_v<double>);
+    template <>
+    inline constexpr long double inv_sqrt3_v<long double> = static_cast<long double>(inv_sqrt3_v<double>);
+    template <>
+    inline constexpr long double egamma_v<long double> = static_cast<long double>(egamma_v<double>);
+    template <>
+    inline constexpr long double phi_v<long double> = static_cast<long double>(phi_v<double>);
+#endif // ^^^ !defined(__cpp_lib_concepts) ^^^

```

Comment:
There should be a newline after this `#endif` for the usual reason (different section of code).

Reply:
Wait, so after for 165 but before for 57? What about 84? I am now confused 😕

Reply:
> Wait, so after for 165 but before for 57?

Yep. It's confusing because of the preprocessor directives, but the principle is that there should be a newline between different "chunks" of code. Separately, if we need to have preprocessor directives, we typically don't have any newlines immediately within them - we just start and stop the code. That is:

```cpp
kittens;
cats;

#if _HAS_DOGS
puppies;
dogs;
#endif // _HAS_DOGS

cubs;
bears;
```

If dogs were unconditional, we'd have:

```cpp
kittens;
cats;

puppies;
dogs;

cubs;
bears;
```

For line 84 (the `#else`), we usually don't put newlines before/after, but we can if the surrounding code is very dense and newlines would improve readability.

Reply:
I'll push a change to adjust this formatting (and merge with recent clang-format changes).

Reply:
I believe I understand now. Thanks for the explaination and example :)